bCNC/plugins/flatpath.py

```diff
@@ -27,39 +29,70 @@ def __init__(self, master):
 
 	# ----------------------------------------------------------------------
 	# This method is executed when user presses the plugin execute button
 	# ----------------------------------------------------------------------
 	def execute(self, app):
-		#print("go!")
+		flatmethod = self['Method']
+		name = self['Name']
+		if name == "default" or name == "":	name = None
+		if flatmethod == 'Unwind helical':
+			blocks = self.unwind_helical(app=app, parname=name)
+		elif flatmethod == 'Remove Z parameter':
+			blocks = self.replace_z(app=app, parname=name)
+		elif flatmethod == 'Movements at stock surface':
+			blocks = self.replace_z(app=app, parname=name, zrepl=CNC.fmt('z',CNC.vars['surface']))
+		elif flatmethod == '':
+			app.setStatus(_("Operator Error: FlatPath-Please select flattening method"))  # <<< feed back result
+			return
+		else:
+			app.setStatus(_("Internal Error: FlatPath-Unkown method"))  # <<< feed back result
+			return
+
+		active=-1 #add to end
+		if blocks:
+			app.gcode.insBlocks(active, blocks, "Shape flattened") #<<< insert blocks over active block in the editor
+			app.refresh()                                                                                           #<<< refresh editor
+			app.setStatus(_("Generated: FlatPath"))                           #<<< feed back result
+		else:
+			app.setStatus(_("Nothing Generated: FlatPath (Did you select a work path?)"))                           #<<< feed back result
+
+	def unwind_helical(self, app, parname):
 		blocks  = []
 		for bid in app.editor.getSelectedBlocks():
 			if len(app.gcode.toPath(bid)) < 1: continue
-
-			#nblock = Block("flat "+app.gcode[bid].name())
-			#for i in app.gcode[bid]:
-			#	nblock.append(re.sub(r"\s?z-?[0-9\.]+","",i))
-			#blocks.append(nblock)
-
-			eblock = Block("flat "+app.gcode[bid].name())
+			eblock = Block((parname or 'flat ')+app.gcode[bid].name())
 			eblock = app.gcode.fromPath(app.gcode.toPath(bid)[0],eblock)
 			blocks.append(eblock)
+		return blocks
 
+	def replace_z(self, app, parname, zrepl = None):
+		def zreplacer(matchobj):
+			if matchobj.lastindex == 1:
+				return zrepl
+			return matchobj.string[matchobj.regs[0][0]:matchobj.regs[0][1]]
+		zrepl = zrepl or ''
+		blocks  = []
+		for bid in app.editor.getSelectedBlocks():
+			if len(app.gcode.toPath(bid)) < 1: continue
+			nblock = Block((parname or 'flat ')+app.gcode[bid].name())
+			for gline in app.gcode[bid]:
+				line = self.zparamRegexp.sub(zreplacer,gline)
+				nblock.append(line)
+			blocks.append(nblock)
 
-
-		#active = app.activeBlock()
-		#if active == 0: active+=1
-		active=-1 #add to end
-		app.gcode.insBlocks(active, blocks, "Shape flattened") #<<< insert blocks over active block in the editor
-		app.refresh()                                                                                           #<<< refresh editor
-		app.setStatus(_("Generated: Flat"))                           #<<< feed back result
-		#app.gcode.blocks.append(block)
+		return blocks
```

Comment:
Please put newline at the end of the file.

Reply:
Implemented.